src/uast.cc

```diff
@@ -137,95 +244,155 @@ void *UastIteratorNext(UastIterator *iter) {
-error1:
-  xmlFreeDoc(doc);
+  return result.numberVal;
+}
+
+const char *UastFilterString(const Uast *ctx, void *node, const char *query) {
+  auto result = UastFilterTyped(ctx, node, query);
 
-  if (!ok) {
-    NodesFree(nodes);
+  if (!checkResult(result, XPATHTYPE_STRING, "UastFilterString")) {
     return nullptr;
   }
-  return nodes;
+
+  return result.stringVal;
 }
 
-char *LastError(void) {
-  return strndup(error_message, BUF_SIZE);
+UastTypedResult UastFilterTyped(const Uast *ctx, void *node, const char *query) {
+  UastTypedResult ret;
+
+  try {
+
+    auto handler = static_cast<xmlGenericErrorFunc>(Error);
+    initGenericErrorDefaultFunc(&handler);
+
+    QueryResult qobject(ctx, node, query);
+    ret.type = static_cast<XPathType>(qobject.xpathObj->type);
+
+    switch (qobject.xpathObj->type) {
+      case XPATH_BOOLEAN:
+        ret.boolVal = qobject.xpathObj->boolval;
+        break;
+
+      case XPATH_NUMBER:
+        ret.numberVal = qobject.xpathObj->floatval;
+        break;
+
+      case XPATH_STRING:
+        {
+          char *cstr = reinterpret_cast<char *>(qobject.xpathObj->stringval);
+          if (!cstr) {
+            Error(nullptr, "string query returned null string\n");
+            ret.hasError = true;
+            break;
+          }
+          ret.stringVal = strdup(cstr);
+        }
+        break;
+
+      case XPATH_NODESET:
+        {
+          if (!qobject.xpathObj->nodesetval) {
+            Error(nullptr, "Unable to get array of result nodes");
+            ret.hasError = true;
+            break;
+          }
+
+          try {
+            ret.nodesVal = new Nodes();
+          } catch (const std::bad_alloc&) {
+            Error(nullptr, "Unable to get memory\n");
+            ret.hasError = true;
+            break;
+          }
+          if (!qobject.xpathObj->nodesetval) {
+            Error(nullptr, "Unable to get array of results\n");
+            ret.hasError = true;
+            break;
+          }
+
+          auto results = qobject.xpathObj->nodesetval->nodeTab;
+          auto size = qobject.xpathObj->nodesetval->nodeNr;
+          int realSize = 0;
+
+          for (int i = 0; i < size; i++) {
+            if (results[i] != nullptr && results[i]->_private != nullptr) {
+              ++realSize;
+            }
+          }
+
+          if (NodesSetSize(ret.nodesVal, realSize) != 0) {
+            Error(nullptr, "Unable to set nodes size\n");
+            ret.hasError = true;
+            break;
+          }
+
+          // Populate array of results
+          int nodeIdx = 0;
+          for (int i = 0; i < size; i++) {
+            if (results[i] != nullptr && results[i]->_private != nullptr) {
+              ret.nodesVal->results[nodeIdx++] = results[i]->_private;
+            }
+          }
+        }
+        break;
+
+      default:
+        {
+          char msg[128];
+          snprintf(msg, 128, "Unsupported return type (%d)\n", qobject.xpathObj->type);
```

Comment:
There's no need for this (I'd reuse `BUF_SIZE` anyway if needed). You can directly call `Error(nullptr, "...", result->type())`